mxcube3/routes/Queue.py

```diff
@@ -242,155 +234,222 @@ def queue_load_state():
-                except Exception:
-                    logging.getLogger('HWR').exception('[QUEUE] Queue error executing child entry with id: %s' % elem['QueueId'])
-    else:
-        #  not a sample so execute directly
-        logging.getLogger('HWR').info('[QUEUE] Queue executing entry with id: %s' % node_id)
-        if mxcube.queue.queue_hwobj.is_paused():
-            logging.getLogger('HWR').info('[QUEUE] Cannot execute, queue is paused. Waiting for unpause')
-            msg = {'Signal': 'QueuePaused',
-                   'Message': 'Queue execution paused',
-                   'State': 1}
-            socketio.emit('Queue', msg, namespace='/hwr')
-            mxcube.queue.queue_hwobj.wait_for_pause_event()
-
-        entry._view = Mock()  # associated text deps
-        entry._set_background_color = Mock()  # widget color deps
-        #  parent = int(node.get_parent()._node_id)
-        # this is a TaskGroup, so it is not in the parsed queue
-        parent_node = node.get_parent() 
-        #  go a level up,
-        # this is a TaskGroup for a Char, a sampleQueueEntry if DataCol
-        parent_node = parent_node.get_parent()  
-        if isinstance(parent_node, qmo.TaskGroup):
-            parent_node = parent_node.get_parent()
-        parent = int(parent_node._node_id)
-
-        mxcube.queue.last_queue_node.update({'id': node_id,
-                                             'sample': queue[parent]['SampleId']
-                                             })
-        # mxcube.queue.queue_hwobj.execute_entry = types.MethodType(Utils.my_execute_entry, mxcube.queue.queue_hwobj)
-        mxcube.queue.queue_hwobj.execute_entry(entry)
-        entry.set_enabled(False)
+    mxcube.queue.queue_hwobj.execute_entry(entry)
 
     msg = {'Signal': 'QueueStopped',
            'Message': 'Queue execution stopped',
            'State': 1}
+
     socketio.emit('Queue', msg, namespace='/hwr')
 
     return Response(status=200)
 
 
 @mxcube.route("/mxcube/api/v0.1/queue", methods=['POST'])
-def add_sample():
-    '''
-    Add a sample to the queue.
-        :request Content-Type: application/json, {"SampleId": sampleId},
-        where sampleId is sample location (eg '1:01')
-        :response Content-Type: application/json, {"QueueId": node_id,
-            "SampleId": sampleId}, where sampleId is the same as the
-            caller id (eg '1:01') and node_id an integer which is used
-            for refering to this element from now onwards.
-        :statuscode: 200: no error
-        :statuscode: 409: sample could not be added, possibly because
-            it already exist in the queue
-        :example request:   * POST http://host:port/mxcube/api/v0.1/queue
-                            * Content-Type: application/json
-                            * {"SampleId": "1:07"}
-    '''
-    params = request.data
-    params = json.loads(params)
-    sample_id = params['SampleId']
+def queue_set():
```

Comment:
I see, its strange that you experience such big delays when setting the queue. You are right sending and setting the entire queue is intrinsically a O(n) operation. 

We can perhaps not assume that adding a task is in the order of milliseconds, it should reasonably be less. Adding a task is basically appending an object to a list. I looked into the time it takes to add one data collection task since it seems you experience some unreasonably long delays. I timed it to 0.0007s, so adding 1000 tasks should take 0.7s, excluding the HTTP request.

Also keep in mind that inserting instead of appending n data collections in the queue is also a O(n) operation, plus the time it takes to find where to insert the tiem so that would result in 2*O(n) operation

We have not worked with large number of data collections in MxCuBE3 yet so we have not really noticed the delay, but the delay is there regardless of method used. At some point we have to add large number of data collections in one shot regardless of how its done. Speaking about delays, we don't see it as a big problem if the user have to wait 1-2 seconds before the queue starts. However it could be perturbing for the user if they have to wait while adding tasks.

It is possible that we have some performance issues when adding tasks, its even likely that we can optimize it (0.0007s seems a bit long). We have to look in to that but the fact that sending queue is an O(n) operation should not be seen as a problem.

That being said we of course need to find were this delay that you experience come from, but we need to do that regardless of method used.